### ISSUE TITLE
Primitive not bound for QtGui when running on a headless system (readline)

### ISSUE THREAD

**Issue by sensestage** [supercollider/supercollider#1209]
Supposedly this was fixed in:
https://github.com/supercollider/supercollider/commit/f2597d65842009116abeb8f5ddd564be815b04cd

But I still get the error at startup:
```
SC FFT global init: cosTable initialised.
init_OSC
empty
compiling class library...
    NumPrimitives = 586
    compiling dir: '/usr/local/share/SuperCollider/SCClassLibrary'
    compiling dir: '/usr/local/share/SuperCollider/Extensions'
    pass 1 done
ERROR: Class extension for nonexistent class 'Document'
     In file:'/deprecated/3.7/deprecated-3.7.sc'
    numentries = 814835 / 11167620 = 0.073
    4990 method selectors, 2238 classes
    method table size 6191868 bytes, big table size 44670480
    Number of Symbols 11653
    Byte Code Size 351432
    compiled 316 files in 2.56 seconds
compile done
ERROR: A primitive was not bound. 0 585
Instance of Method {    (0xb61d20, gc=01, fmt=00, flg=11, set=04)
  instance variables [15]
    raw1 : Float 0.000000   00000000 0038000C
    raw2 : Float 0.000000   00000200 02010002
    code : instance of Int8Array (0xb61ee0, size=4, set=3)
    selectors : nil
    constants : nil
    prototypeFrame : instance of Array (0xb61e60, size=2, set=3)
    context : nil
    argNames : instance of SymbolArray (0xb61de0, size=2, set=3)
    varNames : nil
    sourceCode : nil
    ownerClass : class Meta_QtGUI (0xb59220)
    name : Symbol 'prSetStyle'
    primitiveName : Symbol '_Qt_SetStyle'
    filenameSymbol : Symbol
'/usr/local/share/SuperCollider/SCClassLibrary/Common/GUI/Base/QtGUI.sc'
    charPos : Integer 3196
}
WARNING: Qt style initialisation failed
ERROR: A primitive was not bound. 0 585
Instance of Method {    (0xb41360, gc=01, fmt=00, flg=11, set=04)
  instance variables [15]
    raw1 : Float 0.000000   00000000 0038000C
    raw2 : Float 0.000000   00000200 02010002
    code : instance of Int8Array (0xb41520, size=4, set=3)
    selectors : nil
    constants : nil
    prototypeFrame : instance of Array (0xb414a0, size=2, set=3)
    context : nil
    argNames : instance of SymbolArray (0xb41420, size=2, set=3)
    varNames : nil
    sourceCode : nil
    ownerClass : class Meta_Font (0xb3d860)
    name : Symbol 'prDefaultFamilyForStyle'
    primitiveName : Symbol '_QFont_DefaultFamilyForStyle'
    filenameSymbol : Symbol
'/usr/local/share/SuperCollider/SCClassLibrary/Common/GUI/Base/QFont.sc'
    charPos : Integer 1909
}
WARNING: Qt font initialisation failed
LID: event loop started
Class tree inited in 0.15 seconds
Cleaning up temp synthdefs...
ERROR: A primitive was not bound. 0 585
Instance of Method {    (0xb61260, gc=01, fmt=00, flg=11, set=04)
  instance variables [15]
    raw1 : Float 0.000000   00000000 0030000C
    raw2 : Float 0.000000   00000100 01000001
    code : instance of Int8Array (0xb61320, size=4, set=3)
    selectors : nil
    constants : nil
    prototypeFrame : instance of Array (0xb6787ea0, size=1, set=3)
    context : nil
    argNames : instance of SymbolArray (0xb6787f20, size=1, set=3)
    varNames : nil
    sourceCode : nil
    ownerClass : class Meta_QtGUI (0xb59220)
    name : Symbol 'palette'
    primitiveName : Symbol '_Qt_GlobalPalette'
    filenameSymbol : Symbol
'/usr/local/share/SuperCollider/SCClassLibrary/Common/GUI/Base/QtGUI.sc'
    charPos : Integer 2627
}
ERROR: Primitive 'none' failed.
Failed.
RECEIVER:
   nil
CALL STACK:
    MethodError:reportError   0x222bd0
        arg this = <instance of PrimitiveFailedError>
    Nil:handleError   0xdd1c50
        arg this = nil
        arg error = <instance of PrimitiveFailedError>
    Thread:handleError   0x20c0d0
        arg this = <instance of Thread>
        arg error = <instance of PrimitiveFailedError>
    Object:throw   0x24b3b0
        arg this = <instance of PrimitiveFailedError>
    Object:primitiveFailed   0xdecc90
        arg this = nil
    < closed FunctionDef >  (no arguments or variables)
    Function:doOnStartUp   0xe5dc50
        arg this = <instance of Function>
    ArrayedCollection:do   0x353a90
        arg this = [*6]
        arg function = <instance of Function>
        var i = 5
    List:do   0x353c70
        arg this = <instance of List>
        arg function = <instance of Function>
    Meta_StartUp:run   0x355230
        arg this = <instance of Meta_StartUp>
    Main:startup   0x38b810
        arg this = <instance of Main>
        var didWarnOverwrite = false
^^ The preceding error dump is for ERROR: Primitive 'none' failed.
Failed.
RECEIVER: nil
```

**Comment by colinsullivan** [comment]
I am experiencing this as well.  Any pointers for a workaround?

```
init_OSC
empty
compiling class library...
        NumPrimitives = 589
        compiling dir: '/usr/local/share/SuperCollider/SCClassLibrary'
        compiling dir: '/usr/local/share/SuperCollider/Extensions'
        pass 1 done
ERROR: Class extension for nonexistent class 'Document'
     In file:'/deprecated/3.7/deprecated-3.7.sc'
        numentries = 817217 / 11213540 = 0.073
        5015 method selectors, 2236 classes
        method table size 6158844 bytes, big table size 44854160
        Number of Symbols 11663
        Byte Code Size 352615
        compiled 316 files in 1.20 seconds
compile done
ERROR: A primitive was not bound. 0 588
Instance of Method {    (0x2666a60, gc=01, fmt=00, flg=11, set=04)
  instance variables [15]
    raw1 : Float 0.000000   00000000 0038000C
    raw2 : Float 0.000000   00000200 02010002
    code : instance of Int8Array (0x2666c20, size=4, set=3)
    selectors : nil
    constants : nil
    prototypeFrame : instance of Array (0x2666ba0, size=2, set=3)
    context : nil
    argNames : instance of SymbolArray (0x2666b20, size=2, set=3)
    varNames : nil
    sourceCode : nil
    ownerClass : class Meta_QtGUI (0x265e3a0)
    name : Symbol 'prSetStyle'
    primitiveName : Symbol '_Qt_SetStyle'
    filenameSymbol : Symbol '/usr/local/share/SuperCollider/SCClassLibrary/Common/GUI/Base/QtGUI.sc'
    charPos : Integer 3072
}
ERROR: Primitive 'none' failed.
Failed.
RECEIVER:
   nil

PROTECTED CALL STACK:
        Meta_MethodError:new    0x1fc93e0
rg this = PrimitiveFailedError
                arg what = Failed.
                arg receiver = nil
        Meta_PrimitiveFailedError:new   0x1fccce0
                arg this = PrimitiveFailedError
                arg receiver = nil
        Object:primitiveFailed  0x1ed4960
                arg this = nil
        Function:prTry  0x212be60
                arg this = a Function
                var result = nil
                var thread = a Thread
                var next = nil
                var wasInProtectedFunc = false

CALL STACK:
        MethodError:reportError   0x2883170
                arg this = <instance of PrimitiveFailedError>
        Nil:handleError   0x2883230
                arg this = nil
                arg error = <instance of PrimitiveFailedError>
        Thread:handleError   0x2883290
                arg this = <instance of Thread>
                arg error = <instance of PrimitiveFailedError>
        Object:throw   0x2883350
                arg this = <instance of PrimitiveFailedError>
        Function:protect   0x2884cd0
                arg this = <instance of Function>
                arg handler = <instance of Function>
                var result = <instance of PrimitiveFailedError>
        Meta_QtGUI:style_   0x2884df0
                arg this = <instance of Meta_QtGUI>
                arg styleName = "Fusion"
        Meta_QtGUI:initClass   0x28868b0
                arg this = <instance of Meta_QtGUI>
        Meta_Class:initClassTree   0x2886e50
                arg this = <instance of Meta_Class>
                arg aClass = <instance of Meta_QtGUI>
                var implementsInitClass = nil
        ArrayedCollection:do   0x286f1f0
                arg this = [*210]
                arg function = <instance of Function>
                var i = 56
        Meta_Class:initClassTree   0x286e9f0
                arg this = <instance of Meta_Class>
                arg aClass = <instance of Meta_Object>
                var implementsInitClass = nil
        Process:startup   0x286e0b0
                arg this = <instance of Main>
                var time = 1.217886176
        Main:startup   0x286d2d0
                arg this = <instance of Main>
                var didWarnOverwrite = false
^^ The preceding error dump is for ERROR: Primitive 'none' failed.
Failed.
RECEIVER: nil


sc3>
```

**Comment by sensestage** [comment]
workaround is editing the classfile to not call the primitive; so in this
case, not call QtGui::style in QtGui::initClass.

**Comment by vivid-synth** [comment]
@crucialfelix do you mean commenting out the ```this.style = "Fusion";``` line in /usr/local/share/SuperCollider/SCClassLibrary/Common/GUI/Base/QtGUI.sc ? I still see the similar errors after doing that. (This time referring to ```QPalette``` -- if I comment out the ```this.palette = QPalette.light;``` line, I just see more errors.)

**Comment by bagong** [comment]
The (rather radical) workaround is described in the raspberry pi build instructions:

http://supercollider.github.io/development/building-raspberrypi

Step 5 13/14 and *notes*

**Comment by vivid-synth** [comment]
(For posterity: the workaround @bagong suggested is the one that worked for me. This is still an issue as of yesterday's master branch)

**Comment by vivid-synth** [comment]
Is this something that can be fixed for 3.8? I don't want to push back the 3.8 release any longer but it would be very nice to have if it's a simple fix and we're still waiting on Windows changes.

**Comment by vivid-synth** [comment]
If anyone can spend a few mins explaining to me where the fix needs to occur, I'm happy to take a stab at this

**Comment by vivid-synth** [comment]
Since the buil instructions have been changed, and will change in the future, here are the workaround lines:

    sudo mv /usr/local/share/SuperCollider/SCClassLibrary/Common/GUI /usr/local/share/SuperCollider/SCClassLibrary/scide_scqt/GUI
    sudo mv /usr/local/share/SuperCollider/SCClassLibrary/JITLib/GUI /usr/local/share/SuperCollider/SCClassLibrary/scide_scqt/JITLibGUI

**Comment by tiagmoraismorgado** [comment]
definitely something worth fixing as soon as possible

**Comment by vivid-synth** [comment]
My offer still stands to work on this if someone can spend a little while explaining the remaining issues to me

**Comment by brianlheim** [comment]
@vivid-synth AFAICT @timblechmann's suggestion above is a very good solution and easiest to implement. We just add a primitive that returns `true` if SC was built with Qt and `false` if not, and then protect all Qt-dependent primitive calls with a call to that.

I just simulated this by replacing every call to a Qt primitive with something misnamed:

```
[brianheim@BrianMBP SCClassLibrary]$ for file in $(grep -rIl "\t_Q" .); do sed -i "" "s/_Q/_Q_lol_/g" $file; done
```

I eventually got it to boot with no primitive-not-bound errors by commenting out the following lines:

https://github.com/supercollider/supercollider/blob/master/SCClassLibrary/Common/GUI/PlusGUI/Math/PlotView.sc#L15-L19
https://github.com/supercollider/supercollider/blob/master/SCClassLibrary/Common/GUI/Base/QFont.sc#L10-L12
https://github.com/supercollider/supercollider/blob/master/SCClassLibrary/Common/GUI/Base/QtGUI.sc#L8-L9

I also got errors when indexing help files, but I am not sure if that would happen on a system truly built without Qt, because you wouldn't have a help browser at all right? Either way, there are only ~10 loc causing the problem, so I say let's just add a new primitive and excise them.

For anyone else following along, that is a bare bones workaround.

I also learned about this while poking around just now: http://doc.sccode.org/Reference/initClass.html

**Comment by danstowell** [comment]
Brian's/Tim's approach sounds good to me.

**Comment by brianlheim** [comment]
Considering this not closed by #3420; I think this primitive is still a good idea.

**Comment by vivid-synth** [comment]
Sigh, I just tried building from the latest master. The workaround above seems to not be working anymore. During the `make` step:

```
ERROR: A primitive was not bound. 0 580                                                                                                
Instance of Method {    (0x338bb80, gc=01, fmt=00, flg=11, set=04)                                                                     
  instance variables [15]                                                                                                              
    raw1 : Float 0.000000   00000000 0070000C                                                                                          
    raw2 : Float 0.000000   00000200 02010002                                                                                          
    code : instance of Int8Array (0x338be80, size=4, set=2)                                                                            
    selectors : nil                                                                                                                    
    constants : nil                                                                                                                    
    prototypeFrame : instance of Array (0x338bdc0, size=2, set=2)                                                                      
    context : nil                                                                                                                      
    argNames : instance of SymbolArray (0x338bd00, size=2, set=2)                                                                      
    varNames : nil                                                                                                                     
    sourceCode : nil                                                                                                                   
    ownerClass : class Meta_QtGUI (0x337e2c0)                                                                                          
    name : Symbol 'prSetStyle'                                                                                                         
    primitiveName : Symbol '_Qt_SetStyle'                                                                                              
    filenameSymbol : Symbol '/home/air/box_build/supercollider/SCClassLibrary/Common/GUI/Base/QtGUI.sc'                                
    charPos : Integer 2899                                                                                                             
}                                                                                                                                      
ERROR: Primitive '__none' failed.                                                                                                      
Failed.                                                                                                                                
RECEIVER:                                                                                                                              
   nil
```

**Comment by vivid-synth** [comment]
Hm, this seems to work:

(in the `supercollider` directory):

    rm -r ./SCClassLibrary/Common/GUI/
    rm -r ./SCClassLibrary/JITLib/GUI/

**Comment by vivid-synth** [comment]
Yep, still an issue, and workaround does not work.

**Comment by sensestage** [comment]
In the Bela fork we have moved these folders into a `scide_scqt` folder to temporary fix the topic:
https://github.com/BelaPlatform/supercollider/tree/bela_hackery_v02/SCClassLibrary/scide_scqt

But that fork needs manual syncing with master, it would be great if at some point it could be merged into the mainline supercollider, but that can only happen when this issue is fixed.

**Comment by Apteryks** [comment]
Shouldn't we be able to just not installed the Qt* related sclang files, so that they simply wouldn't be compiled/available (which is the point when you don't build sclang with Qt) ? I've started experimenting with a primitive 'System.hasQt', but it's messy, and it will require authors of Qt sclang files to constantly add these skips in their code, which they will understandably forget at times.

**Comment by Apteryks** [comment]
@brianlheim I just pushed to my bugfix branch (currently based on 3.10.0). The last missing bit to write (see the todo: https://github.com/Apteryks/supercollider/commit/a2c5f47b1263875e9c86224c22a51552e5bbf05d#diff-2e85f83e606be148396cff34a3103244R34) is some code to discover if Qt is loadable or not. As this might not be trivial (each platform have their own ways of doing so IIUC), maybe the best approach would be to use a header template which defines a HAS_QT symbol that would be set by CMake at build time.

**Comment by brianlheim** [comment]
oh, you should just use an `#ifdef SC_QT` , we already have something in the build system for that (search PyrPrimitive.cpp for SC_QT).

don't do it in the sc class library, that would be a mess.

**Comment by Apteryks** [comment]
I'm now trying to resolve an error in building the doc (as your had reported in this thread) after disabling the QT related classes:
```
SCDoc: Rendering all documents
SCDoc: Indexing help-files...
SCDoc: Indexed 1350 documents in 4.97 seconds
ERROR: A primitive was not bound. 0 716
Instance of Method {    (0x2f65b00, gc=01, fmt=00, flg=11, set=04)
  instance variables [15]
    raw1 : Float 0.000000   00000000 0060000C
    raw2 : Float 0.000000   00000100 01000001
    code : instance of Int8Array (0x2f65c80, size=4, set=2)
    selectors : nil
    constants : nil
    prototypeFrame : instance of Array (0x2654580, size=1, set=2)
    context : nil
    argNames : instance of SymbolArray (0x2654640, size=1, set=2)
    varNames : nil
    sourceCode : nil
    ownerClass : class Meta_WebView (0x2f61f80)
    name : Symbol 'clearCache'
    primitiveName : Symbol '_QWebView_ClearMemoryCaches'
    filenameSymbol : Symbol '/tmp/guix-build-supercollider-3.10.0.drv-4/SuperCollider-Source/SCClassLibrary/Common/GUI/Base/QWebView.sc'
    charPos : Integer 689
}
ERROR: Primitive '__none' failed.
Failed.
RECEIVER:
   nil
PATH: platform/renderAllHelp.scd

PROTECTED CALL STACK:
        Meta_MethodError:new    0x2e06580
                arg this = PrimitiveFailedError
                arg what = Failed.
                arg receiver = nil
        Meta_PrimitiveFailedError:new   0x2e0cb40
                arg this = PrimitiveFailedError
                arg receiver = nil
        Object:primitiveFailed  0x28d0240
                arg this = nil
        Dictionary:keysValuesArrayDo    0x38c7000
                arg this = IdentityDictionary[ (class HelpBrowser -> a Function) ]
                arg argArray = [ nil, nil, nil, nil, class HelpBrowser, a Function, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil ]
                arg function = a Function
                var i = 4
                var j = 1
                var key = nil
                var val = nil
                var arraySize = nil
        Dictionary:keysValuesDo 0x38b9dc0
                arg this = IdentityDictionary[ (class HelpBrowser -> a Function) ]
                arg function = a Function
        Dictionary:do   0x38baa80
                arg this = IdentityDictionary[ (class HelpBrowser -> a Function) ]
                arg function = a Function
        Meta_NotificationCenter:notify  0x3612380
                arg this = NotificationCenter
                arg object = SCDoc
                arg message = didIndexAllDocs
                arg args = nil
        Meta_SCDoc:indexAllDocuments    0x2d1ee00
                arg this = SCDoc
                arg clearCache = false
                var now = 4.074483018
                var key = Classes/String
                var doc = SCDocEntry("Classes/RingBuffer", "RingBuffer", "(Undocumented class)")
                var nonHelpFiles = [ [ /tmp/guix-build-supercollider-3.10.0.drv-4/SuperCollider-Source/HelpSource/BrokenLink.html, BrokenLink.html ], [ /tmp/guix-build-supercollider-3.10.0.drv-4/SuperCollider-Source/HelpSource/Browse.html, Browse.html ], [ /tmp/guix-build-supercollider-3.10.0.drv-4/SuperCollider-Source/HelpSource/OldHelpWrapper.html, OldHelpWrapper.html ], [ /tmp/guix-build-supercollider-3.10.0.drv-4/SuperCollider-Source/HelpSource/Search.html, Search.html ], [ /tmp/guix-build-supercollider-3.10.0.drv-4/SuperCollider-Source/...etc...
                var undocClasses = IdentitySet[ CosineWarp, QFont, QMultiSliderView, SetBuf, AbstractStepValue, QFileDialog, ScoreStreamPlayer, QDragBoth, VolumeGui, QWebPageAction, SynthDefOld, Pfpar, PmonoArticStream, FreqScopeWindow, PauseStream, SCScope, QLineLayout, QSlider, QToolButtonStyle, TestDependant, ServerShmInterface, QMenuRole, MethodOverride, QStaticText, ContiguousBlock, Spawner, Decorator, OSCMultiResponder, QScrollTopView, Pnsym1, QQuartzComposerView, QRenderProcessTerminationStatus, HIDDeviceDispatcher, PrettyEat, HIDElem...etc...
                var additions = Dictionary[ (Classes/String -> [ /tmp/guix-build-supercollider-3.10.0.drv-4/SuperCollider-Source/HelpSource/Classes/String.ext.schelp ]) ]
        Meta_SCDoc:documents    0x2d32840
                arg this = SCDoc
        Meta_SCDoc:renderAll    0x2d39440
                arg this = SCDoc
                arg includeExtensions = true
        a FunctionDef   0x44efa58
                sourceCode = "SCDoc.helpSourceDir = thisProcess.argv[0];
SCDoc.helpTargetDir = thisProcess.argv[1];
SCDoc.renderAll;
0.exit;

 "
        a FunctionDef   0x2d28dc0
                sourceCode = "<an open Function>"
        Function:prTry  0x30872c0
                arg this = a Function
                var result = nil
                var thread = a Thread
                var next = nil
                var wasInProtectedFunc = false

CALL STACK:
        MethodError:reportError
                arg this = <instance of PrimitiveFailedError>
        Nil:handleError
                arg this = nil
                arg error = <instance of PrimitiveFailedError>
        Thread:handleError
                arg this = <instance of Thread>
                arg error = <instance of PrimitiveFailedError>
        Object:throw
                arg this = <instance of PrimitiveFailedError>
        Function:protect
                arg this = <instance of Function>
                arg handler = <instance of Function>
                var result = <instance of PrimitiveFailedError>
        Interpreter:executeFile
                arg this = <instance of Interpreter>
                arg pathName = "platform/renderAllHelp.scd"
                arg args = [*0]
                var result = nil
                var saveExecutingPath = nil
        Process:interpretCmdLine
                arg this = <instance of Main>
^^ The preceding error dump is for ERROR: Primitive '__none' failed.
Failed.
RECEIVER: nil
```

**Comment by Apteryks** [comment]
Nevermind the previous comment, I could make it work with a simple addition :-).

**Comment by danstowell** [comment]
Congratulations folks :)
